[JIT] Enchance training ops check to be more inclusive and account for possible pybind exceptions

ttanpcs opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81655
* #81777
* #81737
* #81653
* #81562
* __->__ #81561

- Modified is_mutable python binding to accept a string instead of a string_view for better python compatibility.
- Modified argument value adding python bindings to deal with input/self edge case due to inconsistencies in how the first variable is named.
- Modified _is_alias_of and created _contains_alias_of python bindings to accurately find out if values are aliasing, or contain an alias.
- Fixed is_mutable implementation to cover all ops that have mutable optional arguments. (These are all the ops that have the optional arguments 'running_mean' and 'running_var' along with either 'train', 'training' or 'use_input_stats.'

Commit:
[JIT] Enchance training ops check to be more inclusive and account for possible pybind exceptions

[ghstack-poisoned]

Commit:
Update on "[JIT] Enchance training ops check to be more inclusive and account for possible pybind exceptions"




[ghstack-poisoned]

Commit:
Update on "[JIT] Enchance training ops check to be more inclusive and account for possible pybind exceptions"


- Modified is_mutable python binding to accept a string instead of a string_view for better python compatibility.
- Modified argument value adding python bindings to deal with input/self edge case due to inconsistencies in how the first variable is named.
- Modified _is_alias_of and created _contains_alias_of python bindings to accurately find out if values are aliasing, or contain an alias.
- Fixed is_mutable implementation to cover all ops that have mutable optional arguments. (These are all the ops that have the optional arguments 'running_mean' and 'running_var' along with either 'train', 'training' or 'use_input_stats.'

[ghstack-poisoned]

davidberard98 commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -1887,9 +1903,22 @@ void initJITBindings(PyObject* module) {
+        // Thus, this function will return false when it errors out because
+        // only cases where tensors are contained are actually meaningful.
+        try {
+          return toTypeInferredIValue(self).overlaps(
```
Can move this try/catch into python? I imagine there could be sources of `TORCH_INTERNAL_ASSERT` and some of those asserts are unexpected and should be reported to the user (but all of them will appear as c10::Error and return false here). If you make this change then please update the comment as well

goldenxuett replied:
I've deleted this try catch and moved the check into python in the next pr.

davidberard98 commented on torch/csrc/jit/python/init.cpp:
```diff
@@ -1699,8 +1699,15 @@ void initJITBindings(PyObject* module) {
+            // this check ensures that those arguments are checked correctly.
             if (name == "input") {
-              self.addArgumentValue("self", toTypeInferredIValue(value));
+              try {
+                self.addArgumentValue("self", toTypeInferredIValue(value));
+              } catch (const c10::Error& e) {
+                self.addArgumentValue("input", toTypeInferredIValue(value));
+              }
```
nit: would it be possible to instead add a `hasArgumentNamed(std::string)` or similar into `SchemaInfo` so this can be handled with an if/else instead of with exceptions?

Commit:
Update on "[JIT] Enchance training ops check to be more inclusive and account for possible pybind exceptions"


- Modified is_mutable python binding to accept a string instead of a string_view for better python compatibility.
- Modified argument value adding python bindings to deal with input/self edge case due to inconsistencies in how the first variable is named.
- Modified _is_alias_of and created _contains_alias_of python bindings to accurately find out if values are aliasing, or contain an alias.
- Fixed is_mutable implementation to cover all ops that have mutable optional arguments. (These are all the ops that have the optional arguments 'running_mean' and 'running_var' along with either 'train', 'training' or 'use_input_stats.'

[ghstack-poisoned]

Commit:
Update on "[JIT] Enchance training ops check to be more inclusive and account for possible pybind exceptions"


- Modified is_mutable python binding to accept a string instead of a string_view for better python compatibility.
- Modified argument value adding python bindings to deal with input/self edge case due to inconsistencies in how the first variable is named.
- Modified _is_alias_of and created _contains_alias_of python bindings to accurately find out if values are aliasing, or contain an alias.
- Fixed is_mutable implementation to cover all ops that have mutable optional arguments. (These are all the ops that have the optional arguments 'running_mean' and 'running_var' along with either 'train', 'training' or 'use_input_stats.'

[ghstack-poisoned]
